File: QubitServer/src/org/labrad/qubits/resources/PreampBoard.java
package org.labrad.qubits.resources;

import java.util.Map;

import org.labrad.qubits.enums.DacFiberId;
import org.labrad.qubits.enums.DcRackFiberId;

import com.google.common.base.Preconditions;
import com.google.common.collect.Maps;

public class PreampBoard implements BiasBoard {
  private String name;
  private Map<DcRackFiberId, DacBoard> dacBoards = Maps.newHashMap();
  private Map<DcRackFiberId, DacFiberId> dacFibers = Maps.newHashMap();

  public static PreampBoard create(String name) {
    PreampBoard board = new PreampBoard(name);
    return board;
  }

  public PreampBoard(String name) {
    this.name = name;
  }

  public String getName() {
    return name;
  }

  public void setDacBoard(DcRackFiberId channel, DacBoard board, DacFiberId fiber) {
    dacBoards.put(channel, board);
    dacFibers.put(channel, fiber);
  }

  public DacBoard getDacBoard(DcRackFiberId channel) {
    Preconditions.checkArgument(dacBoards.containsKey(channel),
        "No DAC board wired to channel '%s' on board '%s'", channel.toString(), name);
    return dacBoards.get(channel);
  }

  public DacFiberId getFiber(DcRackFiberId channel) {
    Preconditions.checkArgument(dacBoards.containsKey(channel),
        "No DAC board wired to channel '%s' on board '%s'", channel.toString(), name);
    return dacFibers.get(channel);
  }
}

File: QubitServer/src/main/scala/org/labrad/qubits/resources/PreampBoard.scala
package org.labrad.qubits.resources

import scala.collection.mutable

import org.labrad.qubits.enums.DacFiberId
import org.labrad.qubits.enums.DcRackFiberId

object PreampBoard {
  def create(name: String): PreampBoard = {
    new PreampBoard(name)
  }
}

class PreampBoard(val name: String) extends BiasBoard {
  private val dacBoards = mutable.Map.empty[DcRackFiberId, DacBoard]
  private val dacFibers = mutable.Map.empty[DcRackFiberId, DacFiberId]

  def setDacBoard(channel: DcRackFiberId, board: DacBoard, fiber: DacFiberId): Unit = {
    dacBoards.put(channel, board)
    dacFibers.put(channel, fiber)
  }

  def getDacBoard(channel: DcRackFiberId): DacBoard = {
    dacBoards.getOrElse(channel, sys.error(s"No DAC board wired to channel $channel on board $name"))
  }

  def getFiber(channel: DcRackFiberId): DacFiberId = {
    dacFibers.getOrElse(channel, sys.error(s"No DAC board wired to channel $channel on board $name"))
  }
}
